docs/coalton-lisp-interop.md

```diff
@@ -0,0 +1,132 @@
+# Coalton-Lisp Interoperation
+
+## Introduction
+
+Coalton is a language embedded in Lisp, and indeed, Coalton compiles to Lisp code. This document describes what Coalton promises for interoperation.
+
+## Interaction Mode
+
+First and foremost, there are two ways to globally compile Coalton. This is determined by an environment variable `COALTON_ENV` which in turns sets a global variable `*interaction-mode*`. There are two options for this variable:
+
+- `COALTON_ENV=development`, which sets `*interaction-mode*` to `:development`
+- `COALTON_ENV=release`, which sets `*interaction-mode*` to `:release`
+
+As the names imply, these two modes have different behaviors that optimize for either development or release.
+
+**WARNING**: *All* Coalton code, *including* the standard library, *must* be compiled in the same mode. We advise application writers to set the environment variable during their application's build process before Coalton is compiled, and to ensure caches are not stale.
+
+By default, the interaction mode is `:development`.
+
+### Development Mode
+
+Development mode allows most structures to be re-defined. This means:
+
+* Most types turn into CLOS classes.
+* Most types are not flattened or unwrapped in any way.
+* Certain optimizations that obscure debugging are disabled.
+
+### Release Mode
+
+Release mode freezes most structures and optimizes them.
+
+* Most types are defined as frozen `defstruct`s or flattened Lisp data types.
+* Many optimizations are performed.
+
+### Pitfalls of Having Two Modes
+
+Unfortunately, having two modes may mean one inadvertently depend on the behavior in one mode that is not supported in the other. We advise testing your code against both modes.
+
+## Promises of Data and Basic Data Types
+
+**PROMISE**: Every Coalton value exists as a Lisp value. The structure of said value may not always be defined, however, and must be considered an element of type `T` in general circumstances.
+
+**PROMISE**: Coalton `Integer`, `Character`, `String`,  `Single-Float`, and `Double-Float` values correspond to their Lisp counterparts.
+
+**PROMISE**: `COALTON:Boolean` is a Lisp `BOOLEAN`; `COALTON:True` is Lisp `T`, and `COALTON:False` is Lisp `NIL`.
+
+**WARNING**: Lists and other data types may not correspond to Lisp's conception of them.
+
+## Promises of  `define-type`
+
+For the most part, `define-type` will not provide guarantees across interaction modes. One reason to understand why is to consider the type
+
+```
+(define-type Wrapper
+  (Wrap Integer))
+```
+
+Depending on the interaction mode, `Wrap` may actually be similar to the `IDENTITY` function, and `Wrapper` may be representation-equivalent to `Integer`.
+
+Nonetheless, there is one practical guarantee: the existence of Lisp functions for unconstrained constructors. *Unconstrained* means that there are no constraints on any type variables present in the type that describes the value. Consider the following type:
```

Comment:
I think here it might be useful to clarify that you are talking about typeclass constraints, possibly explain how they arise (and how you can check for them using `type-of`) and provide an example of a constrained value which is not allowed.